packages/api/tests/lib/test-granules.js

```diff
@@ -91,3 +282,221 @@ test('getGranuleProductVolume() returns correct product volume', (t) => {
+  const granuleId = randomId('granule');
+  const files = [
+    {
+      bucket: process.env.system_bucket,
+      fileName: `${granuleId}.hdf`,
+      key: randomString(),
+    },
+  ];
+
+  // Create Dynamo granule
+  const newGranule = fakeGranuleFactoryV2({ granuleId: granuleId, status: 'failed' });
+  newGranule.published = false;
+  newGranule.files = files;
+  await granuleModel.create(newGranule);
+
+  // create PG granule
+  const fakePGGranule = fakeGranuleRecordFactory(
+    {
+      granule_id: granuleId,
+      status: 'failed',
+      collection_cumulus_id: t.context.testPgCollection.cumulus_id,
+    }
+  );
+  fakePGGranule.published = false;
+  const [granuleCumulusId] = await granulePgModel.create(t.context.knex, fakePGGranule);
+
+  const newPgGranule = await granulePgModel.get(t.context.knex, { cumulus_id: granuleCumulusId });
+  const newDynamoGranule = await granuleModel.get({ granuleId: newGranule.granuleId });
+
+  await deleteGranuleAndFiles({
+    knex: t.context.knex,
+    dynamoGranule: newDynamoGranule,
+    pgGranule: newPgGranule,
+    granuleModelClient: granuleModel,
+  });
+
+  // Check Dynamo and RDS. The granule should have been removed from both.
+  await t.throwsAsync(
+    granuleModel.get({ granuleId: newDynamoGranule.granuleId }),
+    { instanceOf: RecordDoesNotExist }
+  );
+
+  await t.throwsAsync(
+    granulePgModel.get(t.context.knex, { granule_id: newPgGranule.granule_id }),
+    { instanceOf: RecordDoesNotExist }
+  );
+});
+
+test.serial('deleteGranuleAndFiles() will not delete a granule or its S3 files if the PG file delete fails', async (t) => {
+  const { newPgGranule, newDynamoGranule, files } = await createGranuleAndFiles(
+    t.context.knex,
+    t.context.testPgCollection.cumulus_id,
+    false
+  );
+
+  const mockFileModel = {
+    tableName: 'files',
+    delete: () => {
+      throw new Error('Delete failed');
+    },
+  };
+
+  await t.throwsAsync(
+    deleteGranuleAndFiles({
+      knex: t.context.knex,
+      dynamoGranule: newDynamoGranule,
+      pgGranule: newPgGranule,
+      filePgModel: mockFileModel,
+      granuleModelClient: granuleModel,
+    }),
+    { instanceOf: Error }
+  );
+
+  // granule should still exist in Dynamo and PG
+  t.true(await granulePgModel.exists(t.context.knex, { granule_id: newPgGranule.granule_id }));
+  t.true(await granuleModel.exists({ granuleId: newDynamoGranule.granuleId }));
+
+  /* eslint-disable no-await-in-loop */
+  for (let i = 0; i < files.length; i += 1) {
+    const file = files[i];
+    // file should still exist in S3
+    t.true(await s3ObjectExists({ Bucket: file.bucket, Key: file.key }));
+    // file should still exist in PG
+    t.true(await filePgModel.exists(t.context.knex, { bucket: file.bucket, key: file.key }));
+  }
+  /* eslint-enable no-await-in-loop */
+});
+
+test.serial('deleteGranuleAndFiles() will delete PG and S3 Files if the PG Granule delete fails', async (t) => {
```

Comment:
Ah, this is interesting: I would have expected that if a PG granule delete fails, then the PG file deletes would also rollback. But that isn't happening because of the separate nested transactions for file deletes. 

So it's just a natural outcome of the delete pattern we've chosen, and I don't know if there's much we can do about that

Reply:
Yeah, exactly. There's a compromise here. 

The thinking was that we definitely need the Files table and the Files in S3 to match. We should guard against a situation where a PG _fails_ to delete but the S3 File deletes fine. That would lead to a mismatch between the Files PG table and what's in S3.

The catch is that all of the File deletion needs to happen before all of the Granule deletion because Files have a foreign key pointing to Granules and Postgres won't let you delete a Granule if Files are still around with a foreign key pointing to it. If that wasn't the case we could get away with trying to delete the granule first and bailing early if it fails.

There may be another way to structure this to preserve everything that I just haven't found yet. Happy to discuss.

Reply:
This has been revised based on our discussion. We took S3 file deletes out of the transaction so they operate independently and don't impact any rollbacks. Everything else lives in one transaction rather than a nested pair:

d7230248ee1b906511c141188c56e31173326976

I hate to ask but a sanity check on the tests in the above commit would be nice. I updated the wording and the `exists = true/false` checks based on what we have in the new flow.